File: Overriding.cpp
#include <iostream>

using namespace std;

class Base
{
    public:
        int i , j;

        void fun()
        {
            cout<<"Inside Base fun\n";
        }

        void gun()
        {
            cout<<"Inside  Base gun\n";
        }
        virtual void sun()
        {
            cout<<"Inside Base sun\n";
        }
        virtual void bun()
        {
            cout<<"Inside Base bun\n";
        }

};

class Derived : public Base
{
    public:
        int x;

        void gun()
        {
            cout<<"Inside Derived gun\n";
        }

        void sun()
        {
            cout<<"Inside Derived sun\n";
        }

        void run()
        {
            cout<<"Inside Derived run\n";
        }

        virtual void mun()
        {
            cout<<"Inside Derived mun\n";
        }
};

int main()
{
    Base * bp1 = new Base();
    Base * bp2 = new Derived();

    Derived * dp1 = new Derived();
    //Derived * dp2 = new Base();           // downcasting

    cout<< sizeof(Base)<<"\n";
    cout<< sizeof(Derived)<<"\n";

    bp2 -> fun();
    bp2 -> gun();
    bp2 -> sun();
    //bp2 -> run();                        // not a member of base class
    //bp2 -> mun();                        // not a member of base class
    bp2 -> bun();


    return 0; 
}

File: Overriding.java
class Base
{
    public int i , j;

    public void fun()
    {
        System.out.println("Inside base fun");
    }

    public void gun()
    {
        System.out.println("Inside base  gun");
    }

    public void sun()
    {
        System.out.println("Inside base sun");
    }

    public void bun()
    {
        System.out.println("INside base bun");
    }
}

class Derived extends Base
{
    public int x;

    public void gun()
    {
        System.out.println("Inside Derived gun");
    }

    public void sun()
    {
        System.out.println("Inside Derived sun");
    }

    public void run()
    {
        System.out.println("Inside Derived run");
    }

    public void mun()
    {
        System.out.println("Inside Derived mun");
    }

}

class Overriding
{
    public static void main(String A[])
    {
        Base bp1 = new Base();
        Base bp2 = new Derived();
        
        Derived dp1 = new Derived();

        bp2.fun();
        bp2.gun();
        bp2.sun();
        //bp2.run();
        //bp2.mun();
        bp2.bun();
        
    }
}